### PR TITLE
roms scan / services

### PR DESCRIPTION
**Pull request by marius851000**
I added a dolphin services, who recursively scan a folder for dolphin roms.
You can configure the scan folder in the dolphin runner option.
It is actually only compatible with wbfs and iso wii files.
It also extract the name for the game name, and the game id for the slug.

update : 
I made it a general rom scanner.

### CODE REVIEW

**Comment by sigmaSd** [comment]
Just a suggestion u can use linux file utility so you don't need to rely on extensions,
Just want to say this fn `rom_read_data` is amazing didn't know about these tricks thx for sharing!

File: lutris/util/scanfolder.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    if isinstance(folder, list):
+        folder_to_scan = folder
+    elif isinstance(folder, str):
+        folder_to_scan = [folder]
+    else:
+        raise BaseException
+
+    file_scanned = []
+
+    while len(folder_to_scan) > 0:
```
Comment (PythonicChemist) [inline lutris/util/scanfolder.py]: You can do a recursive walk with os.walk.

File: lutris/util/scanfolder.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
+import os
+from lutris.util.log import logger
+
+def scan_folder(folder):
+    """do a recursive browse of the folder string / list of string.
+    return as a list with all file under this folder."""
+    if isinstance(folder, list):
```
Comment (PythonicChemist) [inline lutris/util/scanfolder.py]: the function shouldn't autocorrect wrong usage of it.

File: lutris/services/rom.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+            if achar != 0:
+                buff += bytes((achar,))
+            number += 1
+        return buff
+
+    def bytes_to_str(byte):
+        """ transform bytes to string with the default codec """
+        return str(byte)[2:-1]
+
+    rom = open(location, "r+")
```
Comment (PythonicChemist) [inline lutris/services/rom.py]: Use a context manager to open files.

File: lutris/services/rom.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    #    for game in pga.get_games_where(installer_slug=INSTALLER_SLUG,
+    #                                    installed=1)
+    #}
+
+    system_config = LutrisConfig()
+    scan_directory = system_config.system_config["rom_directory"]
+    roms = []
+    roms_slug = []
+
+    for scanned_folder in os.walk(scan_directory,followlinks=True):
```
Comment (PythonicChemist) [inline lutris/services/rom.py]: You can directly unpack it:
```
for folder, subfolder, files in os.walk(scan_directory,followlinks=True):
```

File: lutris/services/rom.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+        folder, subfolder, files = scanned_folder
+        for scanned_file in files:
+            element = folder + "/" + scanned_file
+            try:
+                result = rom_read_data(element)
+            except:
+                logger.error("failed to add the rom at %s." % element)
+                result = False
+            if result != False:
```
Comment (PythonicChemist) [inline lutris/services/rom.py]: result may be referenced before assignment.

File: lutris/services/rom.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+    )
+    game_config.raw_game_config.update(config)
+    game_config.save()
+
+
+def rom_read_data(location):
+    """ extract data from the rom location at location.
+    return a dict with "data" and "config", to be applied to a game in Lutris """
+    # TODO: extract the image of the rom
+    def scan_to_00(mm, start):
```
Comment (PythonicChemist) [inline lutris/services/rom.py]: What is the reasoning behind having these two functions nested in rom_read_data?

**Comment by PythonicChemist** [comment]
I haven't had the chance to test your code yet. How well does it work for dolphin so far?

File: lutris/services/rom.py (hunk 6 above)
Comment (marius851000) [reply lutris/services/rom.py]: I've put this here because that made them private ( but maybe put them globally could be useful for other use ).

**Comment by strycore** [comment]
Please rebase your PR against the `next`  branch

**Comment by marius851000** [comment]
Not really sure if I rebased well. I've changed the destination to next, because it added at lot of strange commit else.

**Comment by marius851000** [comment]
when I trie to run lutris after the rebase, it told me : 
`2018-11-12 11:14:27,791: lspci is not available. List of graphics cards not available
2018-11-12 11:14:30,002: Running Lutris 0.5.0-alpha
2018-11-12 11:14:30,048: This version of Gtk doesn't support set_option_context_summary
2018-11-12 11:14:31,295: adding SUPER MARIO GALAXY.
2018-11-12 11:14:31,556: adding Le Royaume de Ga'Hoole.
Traceback (most recent call last):
  File "/home/marius/mylutris/lutris/lutris/gui/application.py", line 209, in do_activate
    self.window = LutrisWindow(application=self)
  File "/home/marius/mylutris/lutris/lutris/gui/lutriswindow.py", line 125, in __init__
    self.view = self.get_view(view_type)
  File "/home/marius/mylutris/lutris/lutris/gui/lutriswindow.py", line 325, in get_view
    return GameListView(self.game_store)
  File "/home/marius/mylutris/lutris/lutris/gui/gameviews.py", line 384, in __init__
    self.set_sort_with_column(COL_LASTPLAYED_TEXT, COL_LASTPLAYED)
AttributeError: 'GameListView' object has no attribute 'set_sort_with_column'
Traceback (most recent call last):
  File "/home/marius/mylutris/lutris/lutris/gui/application.py", line 345, in do_command_line
    self.window.present()
AttributeError: 'NoneType' object has no attribute 'present'
2018-11-12 11:14:31,778: Shutting down Lutris
`

It worked before. It appear to be a cause of the rebase.

**Comment by sigmaSd** [comment]
Its expected the next branch is broken currently

**Comment by sigmaSd** [comment]
If you want to keep testing your pr after the rebase just comment out the lines that causes issues I think there are like 3 or so

**Pull request merged**